### PR TITLE
handleYield_corres

### PR DESCRIPTION
**Pull request by mktnk3**
This PR proves handleYield_corres.

Also includes:
  - corres rules for endTimeslice, chargeBudget, etc.
  - some spec updates for abstract and design specs

### CODE REVIEW

File: proof/refine/ARM/Syscall_R.thy [hunk 1]
```diff
@@ -1636,26 +1636,425 @@ lemma setSchedulerAction_obj_at'[wp]:
...
+
+lemma endTimeslice_corres:
+  "corres dc
+     (invs and valid_list and valid_sched_action and active_sc_valid_refills and valid_release_q
+      and valid_ready_qs and scheduler_act_sane and (\<lambda>s. cte_wp_at is_ep_cap (cur_thread s, tcb_cnode_index 4) s)
+      and cur_sc_active and ct_active and schact_is_rct and current_time_bounded 2)
+     invs'
+     (end_timeslice canTimeout) (endTimeslice canTimeout)"
+  (is "corres _ ?pre ?pre' _ _")
+
```
Comment (corlewis) [inline proof/refine/ARM/Syscall_R.thy:1659]: Is this whitespace intentional?

File: proof/refine/ARM/Syscall_R.thy [hunk 2]
```diff
@@ -1636,26 +1636,425 @@ lemma setSchedulerAction_obj_at'[wp]:
...
+  unfolding refillResetRR_def
+  apply (wpsimp wp: updateSchedContext_invs')
+  apply (intro conjI; clarsimp elim!: live_sc'_ex_cap[OF invs_iflive'])
+  by (fastforce dest!: valid_sc_strengthen[OF invs_valid_objs']
+                 simp: valid_sched_context'_def valid_sched_context_size'_def scBits_simps objBits_simps')
+
+lemma cur_sc_tcb_cross:
+  "\<lbrakk>(s, s') \<in> state_relation; valid_objs s; pspace_aligned s; pspace_distinct s;
+    cur_sc_tcb s; schact_is_rct s\<rbrakk>
+  \<Longrightarrow> obj_at' (\<lambda>sc. scTCB sc = Some (ksCurThread s')) (ksCurSc s') s'"
```
Comment (corlewis) [inline proof/refine/ARM/Syscall_R.thy:1827]: This could probably be moved earlier as well, in case we want to cross it over more often.

**Review by corlewis — APPROVED** [review]
Very nice!

File: proof/refine/ARM/Syscall_R.thy [hunk 3]
```diff
@@ -1623,7 +1623,7 @@ lemma hw_invs'[wp]:
    apply (fastforce simp: ct_in_state'_def)
   apply (rename_tac epCap)
   apply (case_tac epCap; clarsimp?, (solves wpsimp)?)
-   apply (wpsimp wp: getNotification_wp)
+   apply_trace (wpsimp wp: getNotification_wp)
```
[the remark this answers is not in the document]
Comment (corlewis) [reply proof/refine/ARM/Syscall_R.thy:1626]: The Theory Linter should now flag any checked in use of `apply_trace`. I'm just rerunning it for this PR now to confirm that it works, I hope that you don't mind.

[the remark this answers is not in the document]
Comment (mktnk3) [reply proof/refine/ARM/Syscall_R.thy:1626]: Looks like that it is working well :) Are we going to make it as “required”?

[the remark this answers is not in the document]
Comment (corlewis) [reply proof/refine/ARM/Syscall_R.thy:1626]: Yep, looks like it!

I think that we didn't make it required originally due to there being a few cases where these sorts of diagnostic commands are explicitly left in as examples.

File: proof/refine/ARM/Syscall_R.thy [hunk 4]
```diff
@@ -1636,26 +1636,425 @@ lemma setSchedulerAction_obj_at'[wp]:
   unfolding setSchedulerAction_def
   by (wp, clarsimp elim!: obj_at'_pspaceI)
 
+lemma live_sc'_ex_cap:
+  "if_live_then_nonz_cap' s \<Longrightarrow>
+   \<forall>ko. ko_at' ko scPtr s \<longrightarrow> live_sc' ko \<longrightarrow> ex_nonz_cap_to' scPtr s"
```
[the remark this answers is not in the document]
Comment (mktnk3) [reply proof/refine/ARM/Syscall_R.thy:1641]: I am not sure about moving this. This doesn’t add anything new, I just wanted to have the big arrow in the right position for recurring strengthening I needed to do.

**Review by lsf37 — APPROVED** [review]

**Pull request merged**